pep-9999.rst

```diff
@@ -0,0 +1,173 @@
+a custom metaclass, then it is hard to make a corresponding class generic.
+This is particularly hard for library classes that a user doesn't control.
+A workaround is to always mix-in ``GenericMeta``::
+
+  class AdHocMeta(GenericMeta, LibraryMeta):
+      pass
+
+  class UserClass(LibraryBase, Generic[T], metaclass=AdHocMeta):
+      ...
+
+but this is not always practical or even possible. With the help of the
+proposed special attributes the ``GenericMeta`` metaclass will not be needed.
+
+
+Hacks and bugs that will be removed by this proposal:
+-----------------------------------------------------
+
+- ``_generic_new`` hack that exists since ``__init__`` is not called on
+  instances with a type differing form the type whose ``__new__`` was called,
+  ``C[int]().__class__ is C``.
+
+- ``_next_in_mro`` speed hack will be not necessary since subscription will
+  not create new classes.
+
+- Ugly ``sys._getframe`` hack, this one is particularly nasty, since it looks
+  like we can't remove it without changes outside ``typing``.
+
+- Currently generics do dangerous things with private ABC caches
+  to fix large memory consumption that grows at least as O(N\ :sup:`2`),
+  see [2]_. This point is also important because it was recently proposed to
+  re-implement ``ABCMeta`` in C.
+
+- Problems with sharing attributes between subscripted generics,
+  see [3]_. Current solution already uses ``__getattr__`` and ``__setattr__``,
+  but it is still incomplete, and solving this without the current proposal
+  will be hard and will need ``__getattribute__``.
+
+- ``_no_slots_copy`` hack, where we clean-up the class dictionary on every
+  subscription thus allowing generics with ``__slots__``.
+
+- General complexity of the ``typing`` module, the new proposal will not
+  only allow to remove the above mentioned hacks/bugs, but also simplify
+  the implementation, so that it will be easier to maintain.
+
+
+Specification
+=============
+
+Idea of ``__class_getitem__`` is very simple, it is an exact analog of
+``__getitem__`` with an exception that it is called on a class that
+defines it, not on its instances, this allows us to avoid
+``GenericMeta.__getitem__`` for things like ``Iterable[int]``. Note that this
+method is used as a fallback, so if a metaclass defines ``__getitem__``,
+then it will have the priority.
+
+If an object that is not a class object appears in bases of a class
+definition, the ``__subclass_base__`` is searched on it. If found,
+it is given an original tuple of bases as an argument. If the result of call
+is not ``None``, then it is substituted instead of this object.
+Otherwise, the base is just removed. This is necessary to avoid
+inconsistent MRO errors, that are currently prevented by manipulations in
+``GnericMeta.__new__``. After creating the class, original bases are saved
+in ``__orig_bases__`` (now this is also done by the metaclass).
+
+These two method names are reserved for exclusive use by the ``typing`` module
```

Comment:
Prefix with "NOTE:".

Also I'm not sure this restriction is worth it. We can't enforce it, and I don't (yet) want to disallow other uses of annotations. (Maybe in Python 4.0. :-)

Reply:
I could imagine that without this note, the PEP may be perceived as something like ``__init_subclass__`` that is intended to be used to customize classes.